js/kafkaConsumer.js

```diff
@@ -0,0 +1,192 @@
+    }
+
+    async connect() {
+        await this.consumer.connect();
+    }
+
+    async disconnect() {
+        await this.consumer.disconnect();
+        this.running = false;
+    }
+
+    async consume(topic, callback, options = {}) {
+        // if the consumer is already running, stops it to
+        // subscribe to another topic
+        if (this.running) await this.consumer.stop();
+
+        await this.consumer.subscribe({ topic: topic });
+        this.topicCallbacks[topic] = callback;
+
+        // retries processing previously failed messages every second
+        setInterval(() => this._retry(callback, options), 1000);
+
+        this.running = true;
+        this._runConsumer(options);
+    }
+
+    async _runConsumer(options = {}) {
+        this.consumer.run({
+            autoCommit: conf("KAFKA_CONSUMER_AUTO_COMMIT", true),
+            autoCommitInterval: conf("KAFKA_CONSUMER_AUTO_COMMIT_INTERVAL", 5000),
+            autoCommitThreshold: conf("KAFKA_CONSUMER_AUTO_COMMIT_THRESHOLD", 1),
+            partitionsConsumedConcurrently: conf(
+                "KAFKA_CONSUMER_PARTITIONS_CONSUMED_CONCURRENTLY",
+                1
+            ),
+            eachBatchAutoResolve: conf("KAFKA_CONSUMER_BATCH_AUTO_RESOLVE", true),
+            eachBatch: async ({ batch, heartbeat, isRunning, isStale }) => {
+                for (const message of batch.messages) {
+                    // does not process message if message is marked
+                    // as stale or if the consumer is not running
+                    if (!isRunning() || isStale()) break;
+
+                    const parsedMessage = JSON.parse(message.value.toString());
+                    try {
+                        await this.topicCallbacks[batch.topic](parsedMessage);
+                    } catch (err) {
+                        // if the message processing fails, the message is
+                        // added to a retry buffer that will retry in an
+                        // exponentially increasing delay
+                        this.retryBuffer.push({
+                            ...parsedMessage,
+                            firstFailure: Date.now(),
+                            lastRetry: Date.now(),
+                            topic: batch.topic,
+                            retries: conf("KAFKA_CONSUMER_MESSAGE_FAILURE_RETRIES", 5),
+                            retryDelay: conf("KAFKA_CONSUMER_MESSAGE_FIRST_DELAY", 50)
+                        });
+                        this._updateBufferFile();
+                        return;
+                    }
+
+                    await heartbeat();
+
+                    if (!options.autoConfirm) continue;
+                    if (options.onSuccess) options.onSuccess(parsedMessage);
+                    else this._onSuccess(message);
+                }
+            }
+        });
+    }
+
+    async _retry(callback, options = {}) {
+        if (this.retryBuffer.length === 0) return;

```

Comment:
encapsulate this around a `KafkaRetryConsumer` abstraction that inherits from `KafkaConsumer`